### PR TITLE
Configure config drive ISO with password file on reinstallation of VM  with password-enabled template

### LINKED ISSUE #9095: ConfigDrive In L2 Networking Does not offer password In Reinstalling VM
ISSUE TYPE
Bug Report
Improvement Request

COMPONENT NAME
API

CLOUDSTACK VERSION
4.17.2.0

CONFIGURATION
advanced networking

OS / ENVIRONMENT
U20.4 LTS

SUMMARY
I Create Network Offering For L2 Network With "UserData Enable Option". 
I Used ConfigDrive Option In Cloud-init But In New Creation Of VM instance Cloudstack In Mount Drive Presented to me password File But after reinstalling, I don't have that file anymore.

STEPS TO REPRODUC

EXPECTED RESULTS
i want to  use password file that cloudstack Mount for me. in Reinstalling VM
image

ACTUAL RESULTS
i dont see password file 
image

### PR DESCRIPTION
**Pull request by Pearl1594**
Fixes #9095 

Test1:
1. create a VM based off a password enabled template on a L2 network with configdrive support
2. mounted the configdrive iso and verified the password in the password file
3. reinstalled the VM with the same password-enabled template
4. mounted the configdrive iso and verified the password in the password file

Test2:
1. create a VM based off a password enabled template on a L2 network with configdrive support
2. mounted the configdrive iso and verified the password in the password file
3. reinstalled the VM with another template that isn't password enabled
4. mounted the configdrive iso and verified that password file is not present

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by weizhouapache** [comment]
@Pearl1594 
does the issue happen only when restore a running vm ?

when restore a stopped vm, is the password file present in the configdrive ISO ?

**Comment by Pearl1594** [comment]
@weizhouapache on restoring a stopped VM with a password-enabled template, and starting the VM, the password file is present in the configdrive iso. I've updated the test to indicate the same.

**Comment by weizhouapache** [comment]
> @weizhouapache on restoring a stopped VM with a password-enabled template, and starting the VM, the password file is present in the configdrive iso. I've updated the test to indicate the same.

cool, thanks @Pearl1594

**Review by shwstppr — APPROVED** [review]
Code LGTM

**Review by weizhouapache — APPROVED** [review]
code lgtm

need testing

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 1]
```diff
@@ -7844,9 +7844,10 @@ public UserVm restoreVirtualMachine(final Account caller, final long vmId, final
             }
             vm.setPassword(password);
         }
+        UserVmDetailVO userVmPasswordDetail = userVmDetailsDao.findDetail(vm.getId(), VmDetailConstants.PASSWORD);
         if (needRestart) {
             try {
-                if (vm.getDetail(VmDetailConstants.PASSWORD) != null) {
+                if (Objects.nonNull(userVmPasswordDetail)) {
```
Comment (weizhouapache) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:7850]: @Pearl1594 
does this work ?
```
                if (password != null) {
```

**Review by DaanHoogland — COMMENTED** [review]
@Pearl1594 @weizhouapache , can we extract the password logic here and create a call `addPasswordtoDetailsIfNeeded()` (not that I like that name) and make it check and generate. The whole logic is a bit obscure. It seems above the changes a new password is always generated when the template is password enabled, though on a restart you wouldn't alsways want that. And the detail must exist to add it to the vm.details. It may work but that would be kind of magic.

**Comment by weizhouapache** [comment]
> It seems above the changes a new password is always generated when the template is password enabled

@DaanHoogland 
this is expected behavior I think.

if reinstall a running vm, needRestart is set to true
if reinstall a stopped vm, needRestart is set to false

in both cases, the password needs to reset.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java (hunk 1 above)
Comment (Pearl1594) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:7850]: I believe that should work too. I'll check and update for simplicity.

**Comment by Pearl1594** [comment]
@DaanHoogland As Wei mentioned, password reset is needed nonetheless..

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
> @DaanHoogland As Wei mentioned, password reset is needed nonetheless..

yeah, I can imagine it is implemented in several locations now. Each equally hard to read. <ESC>ddinevermind

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Review by borisstoyanov — APPROVED** [review]
LGTM, manually tested it

**Pull request merged**